design/one-pager-fn-claim-conditions.md

```diff
@@ -0,0 +1,255 @@
+- Normal Event (reason: ComposeResources) generated, containing result's 
+  message.
+
+
+#### Setting the Claim's Status
+Currently the only path to communicate a custom message with the user is by 
+defining your own field in the Claim's status.
+For example, we can define an XRD with:
+```yaml
+status:
+  someCommunicationField:
+    - msg: "Something went wrong."
+```
+
+There are a couple issues with this solution.
+- If we need to halt resource reconciliation due to a fatal error, we can do so
+  with the [SDK](https://github.com/crossplane/function-sdk-go)'s
+  `response.Fatal`, however, this does not also allow us to update the XR and
+  Claim for communication with the user.
+- There is an existing field that would be more intuitive to use as it is
+  already performing this same task for Crossplane itself (`status.conditions`).
+
+#### Setting the Composite's Status Conditions
+Currently you can update the Composite's status conditions by setting them with
+SetDesiredCompositeResource.
+There are a couple of limitations to this:
+- it only shows up on the XR
+- it only shows up if there are no fatal results
+
+Example of setting the Composite's status conditions.
+```go
+// includes:
+//    corev1 "k8s.io/api/core/v1"
+//    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+//    xpv1   "github.com/crossplane/crossplane-runtime/apis/common/v1"
+// 
+//    "github.com/crossplane/function-sdk-go/response"
+desiredXR, err := request.GetDesiredCompositeResource(req)
+c := xpv1.Condition{
+    Type:               xpv1.ConditionType("ImageReady"),
+    Status:             corev1.ConditionFalse,
+    LastTransitionTime: metav1.Now(),
+    Reason:             "NotFound",
+    Message:            "The image provided does not exist or you are not "+
+                        "authorized to use it.",
+}
+desiredXR.Resource.SetConditions(c)
+response.SetDesiredCompositeResource(rsp, desiredXR)
+```
+
+## Proposal
+We would like to allow the Composition Function author to:
+- Choose where results go (Claim or XR)
+- Allow results to update the Status Conditions of the XR and Claim
+
+The following sections get into the details of each of the above items.
+
+### Choose Where Results Go
+Currently each result returned by a function will create a corresponding
+event on the XR (if no previous fatal result exist).
+
+We can expand this functionality by allowing the Result to have targets. In
+order to accomplish this, we will need to expand the Result API as follows.
+```protobuf
+message Result {
+  // Omitted for brevity
+
+  Target target = 3;
+}
+
+// Target of Function results.
+enum Target {
+  TARGET_UNSPECIFIED = 0;
+  TARGET_COMPOSITE_ONLY = 1;
+  TARGET_COMPOSITE_AND_CLAIM = 2;
+}
+```
+The reason for having `TARGET_COMPOSITE_AND_CLAIM` and not `TARGET_CLAIM` is an
+implementation limitation. This prevents more involved API changes, and this
+is also consistent with existing behavior (func copies to XR, Crossplane copies
+XR to Claim).
+
```

Comment:
Any thoughts on how the claim reconciler will know which status conditions it should copy from the XR?

Perhaps just a simple array of types in the XR status, e.g.:

```yaml
status:
  # The XR's condition types that should be back-propagated to the claim
  claimConditions: [DatabaseReady, ImageReady]
  # The XR's conditions
  conditions:
  - type: DatabaseReady
    status: True
    reason: Available
  - type: ImageReady
    status: False
    reason: NotFound
    message: The image provided does not exist or you are not authorized to use it.
```

Reply:
If we are fine adding a new field to status, can we just have `claimConditions` hold the actual conditions themselves? This would allow Result.Target to choose Claim directly, instead of CompositeAndClaim.

One benefit of this would be internal errors.
Claim would see:
```yaml
  - type: DatabaseReady
    status: False
    reason: InternalError
    message: "An internal error occurred. If you need immediate assistance, please contact us at _XYZ_."
```

Composite would see:
```yaml
  - type: DatabaseReady
    status: False
    reason: InternalError
    message: "Failed to parse response from AWS: _some_error_"
```

Reply:
Quickly capturing my take from the last functions SIG meeting: I prefer the "mask" approach where we store a set of condition types that also apply to the claim.

The rationale:

* The claim is a user-facing presentation of the XR, so all claim conditions should also apply to the XR.
* Two arrays of conditions on the XR seems easier to confuse - which are true for the XR?

Reply:
Can you add this to the design doc somewhere too, please.